### PR TITLE
[flang][OpenMP] Semantic checks for DOACROSS clause

### PR DESCRIPTION
**Pull request by kparzysz**
Keep track of loop constructs and OpenMP loop constructs that have been entered. Use the information to validate the variables in the SINK loop iteration vector.

### CODE REVIEW

**Comment by kparzysz** [comment]
The predecessor PR: https://github.com/llvm/llvm-project/pull/115396

File: flang/lib/Semantics/check-omp-structure.cpp [hunk 1]
```diff
@@ -933,11 +934,19 @@ void OmpStructureChecker::CheckDistLinear(
...
 
-void OmpStructureChecker::Leave(const parser::OpenMPLoopConstruct &) {
+void OmpStructureChecker::Leave(const parser::OpenMPLoopConstruct &x) {
   if (llvm::omp::allSimdSet.test(GetContext().directive)) {
     ExitDirectiveNest(SIMDNest);
   }
   dirContext_.pop_back();
+
+  assert(!loopStack_.empty() && "Expecting non-empty loop stack");
+  const LoopConstruct &top = loopStack_.back();
```
Comment (tblah) [inline flang/lib/Semantics/check-omp-structure.cpp:944]: nit: I think this variable will be unused on non-assertion builds

File: flang/lib/Semantics/check-omp-structure.cpp [hunk 2]
```diff
@@ -541,6 +541,7 @@ void OmpStructureChecker::Leave(const parser::OpenMPConstruct &) {
 }
 
 void OmpStructureChecker::Enter(const parser::OpenMPLoopConstruct &x) {
+  loopStack_.push_back(&x);
```
Comment (tblah) [inline flang/lib/Semantics/check-omp-structure.cpp:544]: Is this needed for non-assertion builds?

Comment (tblah) [reply flang/lib/Semantics/check-omp-structure.cpp:544]: Thanks for the explanation!

File: flang/lib/Semantics/check-omp-structure.cpp [hunk 3]
```diff
@@ -1656,6 +1664,11 @@ void OmpStructureChecker::ChecksOnOrderedAsStandalone() {
           "Only SINK or SOURCE dependence types are allowed when ORDERED construct is a standalone construct with no ORDERED region"_err_en_US);
     }
   }
+  auto doaClauses = FindClauses(llvm::omp::Clause::OMPC_doacross);
```
Comment (tblah) [inline flang/lib/Semantics/check-omp-structure.cpp:1667]: nit: elsewhere in this file braced initialization is used
```suggestion
  auto doaClauses{FindClauses(llvm::omp::Clause::OMPC_doacross)};
```

See also other cases in this diff

**Review by tblah — APPROVED** [review]
Thanks for this. LGTM with one nit

**Pull request merged**